python/lsst/ci/models.py

```diff
@@ -0,0 +1,177 @@
+        super().__init__(*args, **kwargs)
+        self.toposorted = False
+        self.sorted_groups = []
+
+    def toposort(self) -> Dict[str, Product]:
+        """Topologically sort the product index and return it.
+
+        This mutates this object if it is not already sorted, and it returns
+        itself for convenience.
+
+        Returns
+        -------
+        Dict[str, Product]
+            Topologically sorted Dict (self)
+        """
+        if self.toposorted:
+            return self
+        dep_graph = {name: set(product.dependencies) for name, product in self.items()}
+        topo_sorted_product_lists = tsort.toposort(dep_graph)
+        ordered_product_list = []
+        for group_idx, sort_group in enumerate(topo_sorted_product_lists):
+            self.sorted_groups.append(sort_group)
+            for product_name in sort_group:
+                ordered_product_list.append(self[product_name])
+        self.clear()
+        for product in ordered_product_list:
+            self[product.name] = product
+        self.toposorted = True
+        return self
+
+    def flat_dependencies(self, product: Product, resolved: Optional[Set[Product]] = None) -> List[Product]:
+        """Return and calculate the set of flat dependencies for this product.
+
+        Parameters
+        ----------
+        product
+            The product we are calculating the flat dependencies for.
+        resolved
+            A set which holds dependencies which we have have previously
+            processed, and which we will add to as we process.
+
+        Returns
+        -------
+            A list of the resolved set of dependencies.
+
+        """
+        if resolved is None:
+            resolved = set()
+        for dependency_name in product.dependencies:
+            dependency_product = self[dependency_name]
+            if dependency_product not in resolved:
+                resolved.add(dependency_product)
+                self.flat_dependencies(dependency_product, resolved)
+        return list(resolved)
+
+    def __setitem__(self, key, value):
+        # invalidate the toposort if an item is set
+        self.toposorted = False
+        self.sorted_groups = []
+        super(ProductIndex, self).__setitem__(key, value)
+
+
+@dataclass
+class Product:
+    """Class representing an EUPS product to be built.
+
+    Parameters
+    ----------
+    name
+        The name of the product
+    sha1
+        The sha1 of the product
+    version
+        The version of the product (from VersionDb)
+    dependencies
+        A list of declared dependencies
+    optional_dependencies
+        A list of optional declared dependencies
+    ref
+        Used in prepare for additional information on what was checked out
+    """
+
+    name: str
+    sha1: str
+    version: str
+    dependencies: List[str]
+    optional_dependencies: Optional[List[str]] = None
+    ref: Optional[Ref] = None
+
+
+@dataclass
+class Ref:
```

Comment:
What is a "git ref"? I think you mean "git reference" and if so, it would be helpful to use the full name instead of the abbreviation in the docs.

Please consider calling the class GitReference as well.

Reply:
In this particular case, the git CLI, documentation, and on-disk structure and [github API](https://developer.github.com/v3/git/refs/) (which was the original motivator for this class) actually refer to "ref" or "refs" more frequently, so I think `Ref` is a bit clearer, though I've clarified the origin.